### ISSUE TITLE
pkg/test/world: "text file busy" error

### ISSUE THREAD

**Issue by mpl** [perkeep/perkeep#1086]
Shamelessly copying the commit message in the CL I had proposed for this fix (https://camlistore-review.googlesource.com/c/camlistore/+/9326):

When running devcam test, or even go test ./pkg/... , we would often hit
the "text file busy" error, meaning that the World used in an
integration test, when building itself with 'go run make.go', failed to
copy $CAMLIROOT/tmp/build-gopath/bin/camlistored to
$CAMLIROOT/bin/camlistored.
This happened because a process started from $CAMLIROOT/bin/camlistored
was still running.

The ongoing assumption was the above process had been started from a
World used in a previous test, or even in the current test, and had
somehow failed to stop, or was taking a bit long at doing so.

Therefore, the first fix attempt
(https://camlistore-review.googlesource.com/#/c/9326/1) was to retry
building the World when we detected that the cause of the build failure
was a syscall.ETXTBSY ()error, in case
camlistored was just taking a bit long to die.

Instead of actually helping, this fix seemed to introduce
even weirder errors: some copy(dest, src) operations in make.go,
typically to install a file (such as camlistored, or the gopherjs
output) at its final location would then sometimes fail because the src
was missing. Which seemed impossible since no error was detected when
the src was initially created.

This lead to the following reasoning:
1) Any test/process using a World is building itself in the same,
unique, $CAMLIROOT/tmp/build-gopath/
2) go test, (or devcam test, which is just go test) runs tests
concurrently by default, up to the number of cores on the machine
3) Therefore it can happen that two tests are building a World at the
same time, on the same location. Creating/Removing files under each
other's feet.
Corollary:
4) It can also happen that a World using test is trying to build
camlistored, while another concurrent World using test is right in the
middle of running the same camlistored. Which explains our initial
symptom.

One easy solution is therefore to make sure that none of the integration
tests run concurently, which can be achieved, at least in devcam, by
adding "-p 1" to the go test command. But it seems like a pretty drastic
measure, and it wouldn't fix the problem for running the tests manually,
where it would be up to the user to think of running them with -p 1.

The other solution, implemented by this change, is to make each World
independent of the others, which basically requires each of them to have
their own GOPATH, set in a random temp dir. This is facilitated with the
introduction of the -vgopath flag in make.go.

The drawback of this approach is the huge hit in performance, as not
only we have to recreate a new copy of the source code for each test,
but we also have to rebuild the World (camlistored and others) for each
of these copies.
In order to soften that hit, the -reusepkg in make.go allows us to copy
all the package archives and binaries that were produced in the default
virtual gopath by a previous build, to the gopath of the World. This
makes building each new World much faster. But it means that in order to
benefit from that optimization, one must have already populated the
default virtual gopath (by running make.go). Which is why for devcam
test, we now automatically run make.go once before running all the tests
(and we introduced the related "prebuild" flag). When running the tests
manually, running make.go beforehand is then up to the user.

In addition to solving the concurrency problem for the Worlds, these new
uses of make.go allowed us to get rid of the useGoPath mode (and the
CAMLI_MAKE_USEGOPATH var that goes with it), which was a bit of a wart.

But this all comes at a cost:

at rev 1da0774ab916f24c1f1d92bc67aef1bd1fbfcca7:
% go test ./pkg/test/integration/
ok  	camlistore.org/pkg/test/integration	43.729s
% go test -p 1 ./pkg/test/integration/
ok  	camlistore.org/pkg/test/integration	7.152s

at this change:
% go test ./pkg/test/integration
ok  	camlistore.org/pkg/test/integration	220.365s

Also note that somewhat surprisingly, removing all concurrency for the
integration tests not only circumvents the problem, but makes running
all the tests faster, at least on my laptop. (7s VS 43s).

TL;DR (this change is option 2):

-removing concurrency for tests to circumvent the problem sounds like a
bummer at first, but is trivial to implement, and is actually not a
performance hit in practice.

-fixing the World concurrency problem by giving them their own gopath is
not too hard in itself, but is very costly since it involves copying all
source code everytime, and having to rebuild the World. Even with some
caching we take a big hit. However, we get concurrency correctness, and
nicer modes/options in make.go.

**Comment by mpl** [comment]
https://camlistore-review.googlesource.com/c/camlistore/+/9326 was my initial proposal, but I think we can revisit this issue, once #1004 is fixed , because it affects the root cause.

**Comment by mpl** [comment]
I think this whole issue got fixed as a side-effect of not using a tmp dir as a tmp GOPATH anymore (f8ef262ff965b6ac083af6e2ff7f76823a088260), thanks to Go 1.10 improved caching.
